tiledb/sm/metadata/metadata.cc

```
@@ -0,0 +1,279 @@
+      offset = sizeof(char);
+      std::memcpy(value_vec.data() + offset, &value_type, sizeof(char));
+      offset += sizeof(char);
+      std::memcpy(value_vec.data() + offset, &value_num, sizeof(value_num));
+      offset += sizeof(value_num);
+      std::memcpy(value_vec.data() + offset, value.data(), value_len);
+
+      // Insert to metadata
+      metadata_map_.emplace(
+          std::make_pair(std::string(key), std::move(value_vec)));
+    }
+  }
+
+  // Create metadata index for fast lookups from index
+  metadata_index_.resize(metadata_map_.size());
+  size_t i = 0;
+  for (auto& m : metadata_map_)
+    metadata_index_[i++] = std::make_pair(&(m.first), &(m.second));
+
+  // Note: `metadata_map_` and `metadata_index_` are immutable after this point
+
+  return Status::Ok();
+}
+
+Status Metadata::serialize(Buffer* buff) const {
+  // Do nothing if there are no metadata to serialize
+  if (metadata_map_.empty())
+    return Status::Ok();
+
+  for (const auto& meta : metadata_map_) {
+    auto key_len = (uint32_t)meta.first.size();
+    RETURN_NOT_OK(buff->write(&key_len, sizeof(uint32_t)));
+    RETURN_NOT_OK(buff->write(meta.first.data(), meta.first.size()));
+    RETURN_NOT_OK(buff->write(meta.second.data(), meta.second.size()));
+  }
+
+  return Status::Ok();
+}
+
+const std::pair<uint64_t, uint64_t>& Metadata::timestamp_range() const {
+  return timestamp_range_;
+}
+
+Status Metadata::del(const char* key) {
+  assert(key != nullptr);
+
+  std::unique_lock<std::mutex> lck(mtx_);
+
+  std::vector<uint8_t> value_vec;
+  value_vec.resize(sizeof(char));
+  char del = 1;
+  std::memcpy(value_vec.data(), &del, sizeof(char));
+  metadata_map_.emplace(std::make_pair(std::string(key), std::move(value_vec)));
+
+  return Status::Ok();
+}
+
+Status Metadata::put(
+    const char* key,
+    Datatype value_type,
+    uint32_t value_num,
+    const void* value) {
+  assert(key != nullptr);
+  assert(value_type != Datatype::ANY);
```

If this is a runtime error, better to return an error status than an assertion (assertions are disabled in release builds)

---

These checks are handled in the caller function in `Array`. Those were mostly dev asserts.